### PR TITLE
[Example] JKNet

### PR DESCRIPTION
**Pull request by xnuohz**
## Description
Pytorch Implementation of JKNet in paper [Representation Learning on Graphs with Jumping Knowledge Networks](https://arxiv.org/abs/1806.03536) @mufeili

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR
- [x] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: examples/pytorch/jknet/README.md [hunk 1]
```diff
@@ -0,0 +1,83 @@
+# DGL Implementation of JKNet
+
+This DGL example implements the GNN model proposed in the paper [Representation Learning on Graphs with Jumping Knowledge Networks](https://arxiv.org/abs/1806.03536).
+
+Contributor: [xnuohz](https://github.com/xnuohz)
+
```
Comment (mufeili) [inline examples/pytorch/jknet/README.md:6]: Did the authors release the code? If so, add a link to their repo.

File: examples/pytorch/jknet/README.md [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+| :-: | :-: | :-: | :-: |
+| Metrics(Table 2) | 89.6±0.5 | 89.1±1.1 | 85.8±1.0 |
+| Metrics(DGL) | 86.1±1.5 | 85.1±1.6 | 84.2±1.6 |
+
+* Citeseer
+
+|  | JK-Maxpool | JK-Concat | JK-LSTM |
+| :-: | :-: | :-: | :-: |
+| Metrics(Table 2) | 77.7±0.5 | 78.3±0.8 | 74.7±0.9 |
+| Metrics(DGL) | 70.9±1.9 | 73.0±1.5 | 69.0±1.7 |
```
Comment (mufeili) [inline examples/pytorch/jknet/README.md:83]: If the authors released their code, can you also list the numbers you got from running their code?

File: examples/pytorch/jknet/README.md [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+--gpu              int     GPU index.                          Default is -1, using CPU.
+```
+
+###### Model options
+```
+--epochs           int     Number of training epochs.          Default is 500.
+--lr               float   Adam optimizer learning rate.       Default is 0.01.
+--lamb             float   L2 regularization coefficient.      Default is 0.0005.
+--hid-dim          int     Hidden layer dimensionalities.      Default is 32.
+--num-layers       int     Number of T.                        Default is 1.
```
Comment (mufeili) [inline examples/pytorch/jknet/README.md:46]: "1" -> "5" based on `main.py`.

File: examples/pytorch/jknet/main.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    parser.add_argument("--num-layers", type=int, default=5, help='Number of GCN layers.')
+    parser.add_argument("--mode", type=str, default='cat', help='Type of aggregation.')
+    parser.add_argument("--dropout", type=float, default=0.5, help='Dropout applied at all layers.')
+
+    args = parser.parse_args()
+    print(args)
+
+    acc_lists = []
+
+    for _ in range(10):
```
Comment (mufeili) [inline examples/pytorch/jknet/main.py:129]: Better to make 10 an additional argument

File: examples/pytorch/jknet/main.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+        acc_lists.append(main(args))
+
+    mean = np.around(np.mean(acc_lists, axis=0), decimals=3)
+    std = np.around(np.std(acc_lists, axis=0), decimals=3)
+    print('Total acc: ', acc_lists)
```
Comment (mufeili) [inline examples/pytorch/jknet/main.py:134]: "Total" -> "total" for consistency

File: examples/pytorch/jknet/main.py [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+    # cuda params
+    parser.add_argument('--gpu', type=int, default=-1, help='GPU index. Default: -1, using CPU.')
+    # training params
+    parser.add_argument('--epochs', type=int, default=500, help='Training epochs.')
+    parser.add_argument('--lr', type=float, default=0.005, help='Learning rate.')
+    parser.add_argument('--lamb', type=float, default=0.0005, help='L2 reg.')
+    # model params
+    parser.add_argument("--hid-dim", type=int, default=32, help='Hidden layer dimensionalities.')
+    parser.add_argument("--num-layers", type=int, default=5, help='Number of GCN layers.')
+    parser.add_argument("--mode", type=str, default='cat', help='Type of aggregation.')
```
Comment (mufeili) [inline examples/pytorch/jknet/main.py:121]: list the options with `choices`

File: examples/pytorch/jknet/model.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    def __init__(self,
+                 in_dim,
+                 hid_dim,
+                 out_dim,
+                 num_layers=1,
+                 mode='cat',
+                 dropout=0.):
+        super(JKNet, self).__init__()
+        
+        self.mode = mode.lower()
```
Comment (mufeili) [inline examples/pytorch/jknet/model.py:17]: Probably not needed

File: examples/pytorch/jknet/main.py [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+                  mode=args.mode,
+                  dropout=args.dropout).to(device)
+    
+    best_model = copy.deepcopy(model)
+
+    # Step 3: Create training components ===================================================== #
+    loss_fn = nn.CrossEntropyLoss()
+    opt = optim.Adam(model.parameters(), lr=args.lr, weight_decay=args.lamb)
+
+    # Step 4: training epoches =============================================================== #
```
Comment (mufeili) [inline examples/pytorch/jknet/main.py:63]: "epoches" -> "epochs"

File: examples/pytorch/jknet/README.md (hunk 1 above)
Comment (xnuohz) [reply examples/pytorch/jknet/README.md:6]: I am not sure whether [this](https://github.com/ShinKyuY/Representation_Learning_on_Graphs_with_Jumping_Knowledge_Networks) is their code.

Comment (mufeili) [reply examples/pytorch/jknet/README.md:6]: There is a code link in the homepage of the first author: https://people.csail.mit.edu/keyulux/. He did not release a complete example, but we can still check for the model implementations.

**Comment by mufeili** [comment]
Have you tried tuning the hyperparameters? Did you use the ones from the paper?

File: examples/pytorch/jknet/model.py [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+                 out_dim,
+                 num_layers=1,
+                 mode='cat',
+                 dropout=0.):
+        super(JKNet, self).__init__()
+        
+        self.mode = mode
+        self.dropout = nn.Dropout(dropout)
+        self.layers = nn.ModuleList()
+        self.layers.append(GraphConv(in_dim, hid_dim, activation=F.relu))
```
Comment (mufeili) [inline examples/pytorch/jknet/model.py:20]: Have you tried different options of `norm` for `GraphConv`?

Comment (xnuohz) [reply examples/pytorch/jknet/model.py:20]: The results were not much different when `norm` was set as 'both' or 'right', but became worse when 'none'.

**Comment by xnuohz** [comment]
> Have you tried tuning the hyperparameters? Did you use the ones from the paper?

I just follow the model architectures from the original paper.

**Comment by mufeili** [comment]
> > Have you tried tuning the hyperparameters? Did you use the ones from the paper?
> 
> I just follow the model architectures from the original paper.

How about the hyperparameters like learning rate?

**Comment by xnuohz** [comment]
> > > Have you tried tuning the hyperparameters? Did you use the ones from the paper?
> > 
> > 
> > I just follow the model architectures from the original paper.
> 
> How about the hyperparameters like learning rate?

I tried lr, dropout, lamb and hid-dim, it's useless.

**Comment by mufeili** [comment]
Have you tried tuning the number of GNN layers? The paper said that they tried 1-6. Also 500 epochs might be too much for these two datasets. Early stopping based on validation accuracy does not work well with these two datasets.

**Comment by xnuohz** [comment]
> Have you tried tuning the number of GNN layers? The paper said that they tried 1-6. Also 500 epochs might be too much for these two datasets. Early stopping based on validation accuracy does not work well with these two datasets.
> 
> There's also a comment left to address.

Yes, I've already tried. I set epochs smaller with different GNN layer and mode, it still doesn't work well.

**Review by mufeili — APPROVED** [review]
